### ISSUE TITLE
"Incorrect username or password" when calling Auth.signIn from jest automated test. Auth.signIn succeeds when called from UI or HTTP call

### ISSUE THREAD

**Issue by mogarick** [aws-amplify/amplify-js#6552]
**NOTE: See my comment from Aug 27th 2020 for an update. It's been narrowed to the Auth.signIn method.**

**Describe the bug**
Executing an automated test in jest confirming a new user via `adminConfirmSignUp` and then calling `Auth.signIn` fails with "Incorrect username or password" but if the same user and pass are used via HTTP authentication endpoint or application UI the signin succeeds.

**To Reproduce**

1. Create a user using `Auth` `signUp`. An email is sent with a confirmation code.
2. As this is a test, call `CognitoIdentityServiceProvider` `adminConfirmSignUp` and `adminUpdateUserAttributes` to confirm user and email respectively.
3. Call `Auth``signIn`
4. an `Incorrect username or password.` error message is received.

**Expected behavior**
I expect the user to be correctly signed in as the user, password, userpoolId, region, etc are correct. This is verified because the user is correctly added to Cognito users pool, `Account status` is `CONFIRMED` and the `Email verified` is also correctly set to `true`. 

**Code Snippet**
``` javascript
try {
    const userIdFragment = shortid.generate();
    const username = `<real_emailuser>+${userIdFragment}@anemail.com`;
    const password = "password20";

    const signedUpUser = await Auth.signUp({
      username,
      password,
      attributes: {
        email: username,
        name: "Test",
        family_name: "Test",
      },
    });

    const credentials = new SharedIniFileCredentials({
      profile: "<my-amplify-profile>",
    });
    const cognitoIdentityServiceProvider = new CognitoIdentityServiceProvider({
      accessKeyId: credentials.accessKeyId,
      secretAccessKey: credentials.secretAccessKey,
      region: config.aws_cognito_region,
    });
    const params = {
      UserPoolId: config.aws_user_pools_id,
      Username: username,
    };

    const result = await cognitoIdentityServiceProvider.adminConfirmSignUp(params).promise();

    const params2 = {
      UserAttributes: [
        {
          Name: "email_verified",
          Value: "true",
        },
      ],
      UserPoolId: config.aws_user_pools_id,
      Username: username,
    };
    const result2 = await cognitoIdentityServiceProvider
      .adminUpdateUserAttributes(params2)
      .promise();

    lg(`Signing In ${username} wit password ${password}...`);
    const user = await Auth.signIn(username, password);
    expect(user).not.toBeNull();
  } catch (error) {
    throw error;
  }
```
**Screenshots**
NA

**What is Configured?**
If applicable, please provide what is configured for Amplify CLI:
* Which steps did you follow via Amplify CLI when configuring your resources.
Resources were configured in varios steps. This is the current `amplify status` output:
```
| Category | Resource name                  | Operation | Provider plugin   |
| -------- | ------------------------------ | --------- | ----------------- |
| Auth     | <myapp>Auth                   | No Change | awscloudformation |
| Auth     | userPoolGroups                 | No Change | awscloudformation |
| Api      | <myapp>api                    | No Change | awscloudformation |
| Api      | AdminQueries                   | No Change | awscloudformation |
| Function | myappAuthPostConfirmation   | No Change | awscloudformation |
| Function | myappAuthPreTokenGeneration | No Change | awscloudformation |
| Function | AdminQueriesxxxxxx           | No Change | awscloudformation |
| Function | myappAuthPreSignup          | No Change | awscloudformation |
```
* Which resources do you have configured?
  * `aws-exports` file:
 ```
   const awsmobile = {
    "aws_project_region": "us-west-2",
    "aws_cognito_identity_pool_id": "us-west-2:xxxxxxxxxxxxx",
    "aws_cognito_region": "us-west-2",
    "aws_user_pools_id": "us-west-2_xxxxxxxxx",
    "aws_user_pools_web_client_id": "xxxxxxxxxxx",
    "oauth": {},
    "aws_appsync_graphqlEndpoint": "https://xxxxxxxxxxx.appsync-api.us-west-2.amazonaws.com/graphql",
    "aws_appsync_region": "us-west-2",
    "aws_appsync_authenticationType": "AMAZON_COGNITO_USER_POOLS",
    "aws_cloud_logic_custom": [
        {
            "name": "AdminQueries",
            "endpoint": "https://xxxxxxxxxx.us-west-2.amazonaws.com/develop",
            "region": "us-west-2"
        }
    ]
};
export default awsmobile;
   ```

<details>
  <summary><strong>Environment</strong></summary>

```
npx envinfo --system --binaries --browsers --npmPackages --npmGlobalPackages

System:
    OS: macOS 10.15.6
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
    Memory: 2.03 GB / 32.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 12.16.3 - ~/n/bin/node
    Yarn: 1.21.1 - ~/.yarn/bin/yarn
    npm: 6.14.7 - ~/n/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Browsers:
    Brave Browser: 80.1.5.123
    Chrome: 84.0.4147.105
    Firefox: 72.0.2
    Safari: 13.1.2
  npmPackages:
    @aws-amplify/api: ^3.1.9 => 3.1.21 
    @aws-amplify/pubsub: ^3.0.10 => 3.0.22 
    @babel/core: ^7.6.4 => 7.10.5 
    @carimus/metro-symlinked-deps: ^1.1.0 => 1.1.0 
    @react-native-community/datetimepicker: 2.2.2 => 2.2.2 
    @react-native-community/netinfo: 5.5.1 => 5.5.1 
    @types/jest: ^26.0.7 => 26.0.7 
    @types/node-fetch: ^2.5.7 => 2.5.7 
    @types/object-path: ^0.11.0 => 0.11.0 
    @types/react: ^16.9.11 => 16.9.43 
    @types/react-native: ^0.60.22 => 0.60.31 
    @types/react-native-calendars: ^1.20.7 => 1.20.8 
    @types/react-test-renderer: ^16.9.2 => 16.9.2 
    @types/validatorjs: ^3.15.0 => 3.15.0 
    aws-amplify: ^3.0.10 => 3.0.22 
    aws-amplify-react-native: ^4.1.1 => 4.2.3 
    aws-sdk: ^2.730.0 => 2.730.0 
    babel-loader: ^8.0.6 => 8.1.0 
    babel-plugin-module-resolver: ^4.0.0 => 4.0.0 
    babel-plugin-root-import: ^6.4.1 => 6.5.0 
    babel-plugin-transform-remove-console: ^6.9.4 => 6.9.4 
    babel-preset-expo: ^8.1.0 => 8.1.0 
    debug: ^4.1.1 => 4.1.1 
    docz: ^1.3.2 => 1.3.2 
    docz-theme-default: ^1.2.0 => 1.2.0 
    expo: ^37.0.0 => 37.0.12 
    expo-random: ^8.1.0 => 8.2.1 
    jest-expo: ^37.0.0 => 37.0.0 
    jsonwebtoken: ^8.5.1 => 8.5.1 
    lodash: ^4.17.19 => 4.17.19 
    mobx: ^4.14.1 => 4.15.4 
    mobx-react-lite: ^1.5.0 => 1.5.2 
    mobx-state-tree: ^3.15.0 => 3.17.2 
    nanoid: ^3.1.10 => 3.1.11 
    node-fetch: ^2.6.0 => 2.6.0 
    object-path: ^0.11.4 => 0.11.4 
    react: 16.9.0 => 16.9.0 
    react-art: 16.8.6 => 16.8.6 
    react-dom: 16.9.0 => 16.9.0 
    react-native: https://github.com/expo/react-native/archive/sdk-37.0.0.tar.gz => 0.61.4 
    react-native-calendars: ^1.265.0 => 1.300.0 
    react-native-keyboard-aware-scroll-view: ^0.9.1 => 0.9.1 
    react-native-keyboard-aware-scrollview: ^2.1.0 => 2.1.0 
    react-native-modal-datetime-picker: ^8.5.4 => 8.7.1 
    react-native-paper: ^3.9.0 => 3.11.0 
    react-native-picker-select: ^6.6.0 => 6.6.0 
    react-native-svg-transformer: ^0.14.3 => 0.14.3 
    react-native-web: ^0.11.7 => 0.11.7 
    react-test-renderer: ^16.13.1 => 16.13.1 
    shortid: ^2.2.15 => 2.2.15 
    ts-node: ^8.4.1 => 8.10.2 
    tsconfig-paths: ^3.9.0 => 3.9.0 
    typescript: ^3.8.3 => 3.9.7 
    uuid: ^3.3.3 => 3.4.0 
    validatorjs: ^3.17.1 => 3.18.1 
  npmGlobalPackages:
    @aws-amplify/cli: 4.24.3
    expo-cli: 3.22.3
    juicy-rn: 0.2.7-b7
    lz-string: 1.4.4
    npm: 6.14.7
    ts-node: 8.10.1
    typescript: 3.8.3
    uuid: 7.0.1
```

</details>

**Additional context**
* I even added a 5 second delay after calling confirmation and before calling signIn but I'm getting the same error. 
* Another thing I tried was creating a `PreSignup` Trigger where I set `autoVerifyEmail` and `autoConfirmUser` to true So I didn't need to execute `adminConfirmSignUp` and `adminUpdateUserAttributes`. The result was the same error.
* Using the same user and pass works when I login from UI and also when I get a token via `https://cognito-idp.us-west-2.amazonaws.com/`

**_You can turn on the debug mode to provide more info for us by setting window.LOG_LEVEL = 'DEBUG'; in your app._**

```
console.log node_modules/@aws-amplify/core/lib/Logger/ConsoleLogger.js:109
    [DEBUG] 50:58.693 AuthClass - signUp attrs: [
      { Name: 'email', Value: 'xxxxx+0PiA8tckL@anemail.com' },
      { Name: 'name', Value: 'Test' },
      { Name: 'family_name', Value: 'Test' }
    ]

  console.log node_modules/@aws-amplify/core/lib/Logger/ConsoleLogger.js:109
    [DEBUG] 50:58.714 AuthClass - signUp validation data: null

  console.log node_modules/@aws-amplify/core/lib/Logger/ConsoleLogger.js:109
    [DEBUG] 51:00.377 Hub - Dispatching to auth with  {
      event: 'signUp',
      data: {
        user: CognitoUser {
          username: 'xxxxx+0PiA8tckL@anemail.com',
          pool: [CognitoUserPool],
          Session: null,
          client: [Client],
          signInUserSession: null,
          authenticationFlowType: 'USER_SRP_AUTH',
          storage: [Function],
          keyPrefix: 'CognitoIdentityServiceProvider.xxxxxxxxxxx',
          userDataKey: 'CognitoIdentityServiceProvider.xxxxxxxxxxx.xxxxxx+0PiA8tckL@anemail.com.userData'
        },
        userConfirmed: false,
        userSub: 'xxxxxx-xxxxxx-xxxxx-xxxx-xxxxxxxxxx',
        codeDeliveryDetails: {
          AttributeName: 'email',
          DeliveryMedium: 'EMAIL',
          Destination: 'x***@a***.com'
        }
      },
      message: 'xxxxxxx+0PiA8tckL@anemail.com has signed up successfully'
    }


console.log node_modules/@aws-amplify/core/lib/Logger/ConsoleLogger.js:109
    [DEBUG] 51:15.302 AuthClass - signIn failure {
      code: 'NotAuthorizedException',
      name: 'NotAuthorizedException',
      message: 'Incorrect username or password.'
    }

  console.log node_modules/@aws-amplify/core/lib/Logger/ConsoleLogger.js:109
    [DEBUG] 51:15.303 Hub - Dispatching to auth with  {
      event: 'signIn_failure',
      data: {
        code: 'NotAuthorizedException',
        name: 'NotAuthorizedException',
        message: 'Incorrect username or password.'
      },
      message: 'xxxxxx-xxxxxx-xxxxx-xxxx-xxxxxxxxxx failed to signin'
    }
```

**Comment by mogarick** [comment]
I made another attempt by changing the user creation using `adminCreateUser` and then `Auth` `signIn`. Same result. 
And as my other attempts, using the same created user via UI works just fine as I get directed to the new password view.

It's important to note that for UI I'm using my own UI version not the amplify UI and I'm using the same Auth.xxx calls that I'm trying to make for my automated test so that's why I'm pretty frustrated that I'm getting such errors in the automated test but getting success when signing in via my UI.

Also using the same user and temp pass via HTTP call to `https://cognito-idp.us-west-2.amazonaws.com/` succeeds and I get a JSON response that includes `Challenge_Name=NEW_PASSWORD_REQUIRED` 

``` json
{
  "ChallengeName": "NEW_PASSWORD_REQUIRED",
  "ChallengeParameters": {
    "USER_ID_FOR_SRP": "xxxxxx-xxxx-...",
    "requiredAttributes": "[]",
    "userAttributes": "{\"email_verified\":\"true\",\"email\":\"<the email>\"}"
  },
  "Session": "xxxxxxxxxx..."
}
```
I'm baffled.

Heres is my updated code for this new experiment:

``` javascript
try {
    const credentials = new SharedIniFileCredentials({
      profile: ""<my-amplify-profile>",
    });
    const cognitoIdentityServiceProvider = new CognitoIdentityServiceProvider({
      accessKeyId: credentials.accessKeyId,
      secretAccessKey: credentials.secretAccessKey,
      region: config.aws_cognito_region,
    });

    const userIdFragment = shortid.generate();
    const username = `<real_emailuser>+${userIdFragment}@anemail.com`;
    const password = "password20";
    const passwordTemp = "passwordTemp";

    //Crear nuevo usuario con password temporal.
    const newUserParams = {
      UserPoolId: config.aws_user_pools_id,
      Username: username,
      TemporaryPassword: passwordTemp,
      MessageAction: "SUPPRESS",
      UserAttributes: [
        {
          Name: "email",
          Value: username,
        },
        {
          Name: "email_verified",
          Value: "true",
        },
      ],
    };
    const newUser = await cognitoIdentityServiceProvider.adminCreateUser(newUserParams).promise();
    lg(`Success creating ${username}`);
    lg(JSON.stringify(newUser, null, 2));

    lg(`Signing In ${username} with password ${passwordTemp}...`);
    const user = await Auth.signIn(username, passwordTemp);
    lg("User info");
    lg(user);
    expect(user).not.toBeNull();
  } catch (error) {
    throw error;
  }
```

The errror:

``` json
{
    "code": "NotAuthorizedException",
    "name": "NotAuthorizedException",
    "message": "Incorrect username or password."
}
```

**Comment by mogarick** [comment]
I found a workaround.
I've debugged the amplify code and I noticed that the default `authenticationFlowType` is `USER_SRP_AUTH`. If I add `USER_PASSWORD_AUTH` to the `Auth.config(configParams)`, then the signIn works. Heres is the code snippet:

``` javascript
import config from "~/aws-exports";
Auth.configure({ ...config, authenticationFlowType: "USER_PASSWORD_AUTH" });
...
```
`USER_PASSWORD_AUTH` is the type I use for the HTTP call that works with the same user data that fails with the Jest test because I was receiving a `Missing required parameter SRP_A` error when using `USER_SRP_AUTH`.

As I said this is just a workaround because the ideal is to keep the default `USER_SRP_AUTH` since that's the secure recommendation.

Can someone help me to understand how calling `Auth.signIn` from a Jest test is different from calling it when using the UI (An UI that is not the Amplify one but my own that calls the same `Auth.signIn` method)?

Thank you in advance.

**Comment by khanson98** [comment]
I have a similar issue but it is with CUSTOM_AUTH as well

**Comment by mogarick** [comment]
It looks the problem is definitely related with `Auth.signIn` behavior when called from jest vs called from UI and not from the way a user account is created (either via UI or jest).
If I create and confirm an account using the UI and then I use the same user and password in jest (without creating an account through the jest test), I get the same error `Incorrect username or password` when trying to sign in with `Auth.signIn`. 
So I need to use the same workaround `Auth.configure({ ...config, authenticationFlowType: "USER_PASSWORD_AUTH" });` in order I can successfully Sign in even though the user account I'm using was created via UI.

Another weird thing is that the same call to sign in gets a slightly different response. The only differences (besides the timestamps) are the `username` and the `userDataKey` props. In the jest call the username value includes the email. In the UI call the username value includes the `user_sub` UUID. In this case I'm not sure this is because of the problem or is due to the use of `authenticationFlowType: "USER_PASSWORD_AUTH"` workaround for the jest test.

**Jest Test. Response for `Auth.currentAuthenticatedUser` after an `Auth.signIn` call**
``` json
{
    "username": "<real_emailuser>+p22@anemail.com",
    "pool": {
      "userPoolId": "us-west-<redacted>",
      "clientId": "<redacted>",
      "client": {
        "endpoint": "https://cognito-idp.us-west-2.amazonaws.com/",
        "fetchOptions": {}
      },
      "advancedSecurityDataCollectionFlag": true
    },
    "Session": null,
    "client": {
      "endpoint": "https://cognito-idp.us-west-2.amazonaws.com/",
      "fetchOptions": {}
    },
    "signInUserSession": {
      "idToken": {
        "jwtToken": "<redacted>",
        "payload": {
          "sub": "<user_sub>",
          "email_verified": true,
          "iss": "https://cognito-idp.us-west-2.amazonaws.com/us-west-2<redacted>",
          "cognito:username": "<user_sub>",
          "membershipId": "<redacted>",
          "accountId": "<redacted>",
          "aud": "<redacted>",
          "event_id": "<redacted>",
          "hasManyMemberships": "false",
          "token_use": "id",
          "auth_time": 1598572369,
          "name": "Prueba22",
          "isTheAccountOwner": "true",
          "exp": 1598575971,
          "iat": 1598572371,
          "family_name": "Prueba",
          "groupsEnrolledIn": "[\"<redacted>\",\"<redacted>\",\"<redacted>\"]",
          "email": "<real_emailuser>+p22@anemail.com"
        }
      },
      "refreshToken": {
        "token": "<redacted>"
      },
      "accessToken": {
        "jwtToken": "<redacted>",
        "payload": {
          "sub": "<user_sub>",
          "event_id": "<redacted>",
          "token_use": "access",
          "scope": "aws.cognito.signin.user.admin",
          "auth_time": 1598572369,
          "iss": "https://cognito-idp.us-west-2.amazonaws.com/us-west-2<redacted>",
          "exp": 1598575971,
          "iat": 1598572371,
          "jti": "<redacted>",
          "client_id": "<redacted>",
          "username": "<user_sub>"
        }
      },
      "clockDrift": 0
    },
    "authenticationFlowType": "USER_SRP_AUTH",
    "keyPrefix": "CognitoIdentityServiceProvider.<redacted>",
    "userDataKey": "CognitoIdentityServiceProvider.<redacted>.<real_emailuser>+p22@anemail.com.userData",
    "attributes": {
      "sub": "<user_sub>",
      "email_verified": true,
      "name": "Prueba22",
      "family_name": "Prueba",
      "email": "<real_emailuser>+p22@anemail.com"
    },
    "preferredMFA": "NOMFA"
  }
```

**UI. Response for `Auth.currentAuthenticatedUser` after an `Auth.signIn` call**
``` json
{
  "username": "<user_sub>",
  "pool": {
    "userPoolId": "us-west-2<redacted>",
    "clientId": "<redacted>",
    "client": {
      "endpoint": "https://cognito-idp.us-west-2.amazonaws.com/",
      "fetchOptions": {}
    },
    "advancedSecurityDataCollectionFlag": true
  },
  "Session": null,
  "client": {
    "endpoint": "https://cognito-idp.us-west-2.amazonaws.com/",
    "fetchOptions": {}
  },
  "signInUserSession": {
    "idToken": {
      "jwtToken": "<redacted>",
      "payload": {
        "sub": "<user_sub>",
        "email_verified": true,
        "iss": "https://cognito-idp.us-west-2.amazonaws.com/us-west-2<redacted>",
        "cognito:username": "<user_sub>",
        "membershipId": "<redacted>",
        "accountId": "<redacted>",
        "aud": "<redacted>",
        "event_id": "<redacted>",
        "hasManyMemberships": "false",
        "token_use": "id",
        "auth_time": 1598573323,
        "name": "Prueba22",
        "isTheAccountOwner": "true",
        "exp": 1598576926,
        "iat": 1598573327,
        "family_name": "Prueba",
        "groupsEnrolledIn": "[\"<redacted>\",\"<redacted>\",\"<redacted>\"]",
        "email": "<real_emailuser>+p22@anemail.com"
      }
    },
    "refreshToken": {
      "token": "<redacted>"
    },
    "accessToken": {
      "jwtToken": "<redacted>",
      "payload": {
        "sub": "<user_sub>",
        "event_id": "<redacted>",
        "token_use": "access",
        "scope": "aws.cognito.signin.user.admin",
        "auth_time": 1598573323,
        "iss": "https://cognito-idp.us-west-2.amazonaws.com/us-west-2<redacted>",
        "exp": 1598576926,
        "iat": 1598573327,
        "jti": "<redacted>",
        "client_id": "<redacted>",
        "username": "<user_sub>"
      }
    },
    "clockDrift": 0
  },
  "authenticationFlowType": "USER_SRP_AUTH",
  "keyPrefix": "CognitoIdentityServiceProvider.<redacted>",
  "userDataKey": "CognitoIdentityServiceProvider.<redacted>.<user_sub>.userData",
  "attributes": {
    "sub": "<user_sub>",
    "email_verified": true,
    "name": "Prueba22",
    "family_name": "Prueba",
    "email": "<real_emailuser>+p22@anemail.com"
  },
  "preferredMFA": "NOMFA"
}


```
Here is my updated jest code

``` javascript
Auth.configure({ ...config, authenticationFlowType: "USER_PASSWORD_AUTH" });
  try {
    const username = `<real_emailuser>+p22@anemail.com`;
    const password = "password22";
    await Auth.signIn(username, password);
    const user = await Auth.currentAuthenticatedUser({
      bypassCache: true,
    });
    lg("Signed In User with cache bypassed");
    lg(JSON.stringify(user, null, 2));
    expect(user).not.toBeNull();
  } catch (error) {
    lg(error);
    throw error;
  }
```
### Additional info
* For the UI I'm working with React Native Expo so the bundle gets created using metro. 
* For jest I have configured a `jest.config.js` file with `jest-expo` preset and babel.

**Comment by mogarick** [comment]
Hi again @khanson98 @amhinson. 
I made a new test using now the web version of my app. It worked without any trouble for signing in. 
I could make some parallel tests through my web app to use the same request to make the HTTP call manually. I debugged  `CognitoUser`. I used the `PASSWORD_VERIFIER` server response to the `InitiateAuth`call and then I used the `challenge` value generated for the call to `respondToAuthChallenge`. 
What I found is that apparently the problem is in the `respondToAuthChallenge` data that gets generated when using Jest. 
If I use the same data generated when the call to SignIn is made through the web app / RN iOS app, I can succeed in the call to `RespondToAuthChallenge` when I do it manually using Insomnia (similar to Postman). 
But if I use the data that gets generated when using Jest the call fails in the manual HTTP call.

I looked into the request data to `InitiateAuth`, the server response to that and also the request for the `respondToAuthChallenge ` but I didn't notice any difference. The params are the same. Just with the obvious difference values that are generated after the computing proceses to generate `SRP_A`, and the responses for the challenges received. The SALT received from the server is the same for the 3 examples

Here are the bodies examples
```
//- - - Jest
//InitiateAuth
{
  AuthFlow: "USER_SRP_AUTH",
  ClientId: "<redacted>",
  AuthParameters: {
    USERNAME: "<real_emailuser>+p22@anemail.com",
    SRP_A: "<redacted-JEST>",
  },
  ClientMetadata: {
  },
}
//Server response
{
  ChallengeName: "PASSWORD_VERIFIER",
  ChallengeParameters: {
    SALT: "<redacted>",
    SECRET_BLOCK: "<redacted-JEST>",
    SRP_B: "<redacted-JEST>",
    USERNAME: "<user_sub>",
    USER_ID_FOR_SRP: "<user_sub>",
  },
}

//Response to auth challenge
{
  ChallengeName: "PASSWORD_VERIFIER",
  ClientId: "<redacted>",
  ChallengeResponses: {
    USERNAME: "<user_sub>",
    PASSWORD_CLAIM_SECRET_BLOCK: "<redacted-JEST>",
    TIMESTAMP: "Fri Aug 28 19:22:44 UTC 2020",
    PASSWORD_CLAIM_SIGNATURE: "<redacted-JEST>",
  },
  Session: undefined,
  ClientMetadata: {
  },
}


//- - - UI iOS Device
//Client InitiateAuth
{
  AuthFlow: "USER_SRP_AUTH",
  ClientId: "<redacted>",
  AuthParameters: {
    USERNAME: "<real_emailuser>+p22@anemail.com",
    SRP_A: "<redacted-UIIOSDEVICE>",
  },
  ClientMetadata: {
  },
}

//Server Response
{
  ChallengeName: "PASSWORD_VERIFIER",
  ChallengeParameters: {
    SALT: "<redacted>",
    SECRET_BLOCK: "<redacted-UIIOSDEVICE>",
    SRP_B: "<redacted-UIIOSDEVICE>",
    USERNAME: "<user_sub>",
    USER_ID_FOR_SRP: "<user_sub>",
  },
}

//Client responseToChallenge
{
  ChallengeName: "PASSWORD_VERIFIER",
  ClientId: "<redacted>",
  ChallengeResponses: {
    USERNAME: "<user_sub>",
    PASSWORD_CLAIM_SECRET_BLOCK: "<redacted-UIIOSDEVICE>",
    TIMESTAMP: "Fri Aug 28 19:32:38 UTC 2020",
    PASSWORD_CLAIM_SIGNATURE: "<redacted-UIIOSDEVICE>",
  },
  Session: undefined,
  ClientMetadata: {
  },
}

//- - -UI Chrome web browser
//Client InitiateAuth
{
  AuthFlow: "USER_SRP_AUTH",
  ClientId: "<redacted>",
  AuthParameters: {
    USERNAME: "<real_emailuser>+p22@anemail.com",
    SRP_A: "<redacted-UICHROMEBROWSER>",
  },
  ClientMetadata: {
  },
}

//Server Response
{
  ChallengeName: "PASSWORD_VERIFIER",
  ChallengeParameters: {
    SALT: "<redacted>",
    SECRET_BLOCK: "<redacted-UICHROMEBROWSER>",
    SRP_B: "<redacted-UICHROMEBROWSER>",
    USERNAME: "<user_sub>",
    USER_ID_FOR_SRP: "<user_sub>",
  },
}

//responseToChallenge
{
  ChallengeName: "PASSWORD_VERIFIER",
  ClientId: "<redacted>",
  ChallengeResponses: {
    USERNAME: "<user_sub>",
    PASSWORD_CLAIM_SECRET_BLOCK: "<redacted-UICHROMEBROWSER>",
    TIMESTAMP: "Tue Sep 1 16:58:23 UTC 2020",
    PASSWORD_CLAIM_SIGNATURE: "<redacted-UICHROMEBROWSER>+nRnFoGJ8=",
  },
  Session: undefined,
  ClientMetadata: {
  },
}


```

So at the moment I have this tests results for `Auth.signIn` with `USER_SRP_AUTH`:
* React Native (via expo). OK.
* Web Browser (using React Native Web via expo). OK
* HTTP manual call using the data created from the web browser debugging. OK.
* Jest test. FAIL.
* HTTP manual call using the data created from the Jest test debugging. FAIL.

So I think the problem could be narrowed down to the `CognitoUser.authenticateUserDefaultAuth` code fragment where the `challengeResponses` are created.

I don't have the knowledge to understand well that process. Can Anyone help me or give me a hint about how to dig more into that part?

Here is the  `CognitoUser.authenticateUserDefaultAuth` fragment related.

``` javascript
this.client.request('InitiateAuth', jsonReq, (err, data) => {
				if (err) {
					return callback.onFailure(err);
				}

				const challengeParameters = data.ChallengeParameters;

				this.username = challengeParameters.USER_ID_FOR_SRP;
				serverBValue = new BigInteger(challengeParameters.SRP_B, 16);
				salt = new BigInteger(challengeParameters.SALT, 16);
				this.getCachedDeviceKeyAndPassword();

				authenticationHelper.getPasswordAuthenticationKey(
					this.username,
					authDetails.getPassword(),
					serverBValue,
					salt,
					(errOnHkdf, hkdf) => {
						// getPasswordAuthenticationKey callback start
						if (errOnHkdf) {
							callback.onFailure(errOnHkdf);
						}

						const dateNow = dateHelper.getNowString();

						const message = CryptoJS.lib.WordArray.create(
							Buffer.concat([
								Buffer.from(this.pool.getUserPoolId().split('_')[1], 'utf8'),
								Buffer.from(this.username, 'utf8'),
								Buffer.from(challengeParameters.SECRET_BLOCK, 'base64'),
								Buffer.from(dateNow, 'utf8'),
							])
						);
						const key = CryptoJS.lib.WordArray.create(hkdf);
						const signatureString = Base64.stringify(HmacSHA256(message, key));

						const challengeResponses = {};

						challengeResponses.USERNAME = this.username;
						challengeResponses.PASSWORD_CLAIM_SECRET_BLOCK =
							challengeParameters.SECRET_BLOCK;
						challengeResponses.TIMESTAMP = dateNow;
						challengeResponses.PASSWORD_CLAIM_SIGNATURE = signatureString;

						if (this.deviceKey != null) {
							challengeResponses.DEVICE_KEY = this.deviceKey;
						}

                                                //HERE IS THE CALL TO RESPOND TO THE AUTH CHALLENGE.
						const respondToAuthChallenge = (challenge, challengeCallback) =>

							this.client.request(
								'RespondToAuthChallenge',
								challenge,
								(errChallenge, dataChallenge) => {
									if (
										errChallenge &&
										errChallenge.code === 'ResourceNotFoundException' &&
										errChallenge.message.toLowerCase().indexOf('device') !== -1
									) {
										challengeResponses.DEVICE_KEY = null;
										this.deviceKey = null;
										this.randomPassword = null;
										this.deviceGroupKey = null;
										this.clearCachedDeviceKeyAndPassword();
										return respondToAuthChallenge(challenge, challengeCallback);
									}
									return challengeCallback(errChallenge, dataChallenge);
								}
							);

						const jsonReqResp = {
							ChallengeName: 'PASSWORD_VERIFIER',
							ClientId: this.pool.getClientId(),
							ChallengeResponses: challengeResponses,
							Session: data.Session,
							ClientMetadata: clientMetaData,
						};
						if (this.getUserContextData()) {
							jsonReqResp.UserContextData = this.getUserContextData();
						}
						respondToAuthChallenge(
							jsonReqResp,
							(errAuthenticate, dataAuthenticate) => {
								if (errAuthenticate) {
									return callback.onFailure(errAuthenticate);
								}

								return this.authenticateUserInternal(
									dataAuthenticate,
									authenticationHelper,
									callback
								);
							}
						);
						return undefined;
						// getPasswordAuthenticationKey callback end
					}
				);
				return undefined;
			});
```

**Comment by mogarick** [comment]
I found out the problem is similar to aws/aws-encryption-sdk-javascript#126. The problem is with this code fragment in the `lib-typedarrays.js`

``` javascript
// Handle Uint8Array
	        if (typedArray instanceof Uint8Array) {
	            // Shortcut
	            var typedArrayByteLength = typedArray.byteLength;

	            // Extract bytes
	            var words = [];
	            for (var i = 0; i < typedArrayByteLength; i++) {
	                words[i >>> 2] |= typedArray[i] << (24 - (i % 4) * 8);
	            }

	            // Initialize this word array
	            superInit.call(this, words, typedArrayByteLength);
	        } else {
	            // Else call normal init
	            superInit.apply(this, arguments);
	        }
```

the condition fails presumably due the value coming in from the jest test being generated in other nodejs vm instance or something like that. 

TL;DR I created a custom testEnvironment file as commented by @wi-ski in https://github.com/facebook/jest/issues/7780#issuecomment-615890410 and added `testEnvironment: ".path/to/my/customenv.js",` to the `jest.config.js` file.

Here is the test env file I created.
```
"use strict";

const NodeEnvironment = require("jest-environment-node");

class MyEnvironment extends NodeEnvironment {
  constructor(config) {
    super(
      Object.assign({}, config, {
        globals: Object.assign({}, config.globals, {
          Uint32Array: Uint32Array,
          Uint8Array: Uint8Array,
          ArrayBuffer: ArrayBuffer,
        }),
      })
    );
  }

  async setup() {}

  async teardown() {}
}

module.exports = MyEnvironment;

```
With this tweak I can finally successfully sign in with the default `USER_SRP_AUTH `  authenticationFlowType.

PD: khanson98 may be this can give you a hint for your problem.

PD2: Now I'm facing another one: calling a graphql operation with authMode=AWS_IAM returns this error when executing from Jest. No problem occurs when doing it from the Web/RN App UI. 
```
errors: [ GraphQLError { message: 'Request failed with status code 403' } ]
```

I guess I'll dig-in again and open a new issue if needed.

**Comment by PierrickI3** [comment]
Thanks @mogarick. Your solution worked for me too. I've been scratching my head over this for several days looking at why regular API calls worked for authenticating users while jest reportedly failed with `Incorrect username or password`.

**Comment by bryanrcampbell** [comment]
I was having the same problem with Amplify but was able to get past this by adding `testEnvironment: 'node'` to my jest.config.js as suggesting in a related thread.
